src/daft-io/src/glob.rs

```diff
@@ -0,0 +1,150 @@
+use std::collections::HashSet;
+
+use lazy_static::lazy_static;
+
+lazy_static! {
+    /// Check if a given char is considered a special glob character
+    /// NOTE: we use the `globset` crate which defines the following glob behavior:
+    /// https://docs.rs/globset/latest/globset/index.html#syntax
+    static ref GLOB_SPECIAL_CHARACTERS: HashSet<char> = {
+        let mut set = HashSet::new();
+        set.insert('*');
+        set.insert('?');
+        set.insert('{');
+        set.insert('[');
+        set
+    };
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct GlobFragment {
+    data: String,
```

Comment:
might be a good use of a `Cow` with a lifetime here so you only allocate strings that you have to post-process

Reply:
`data` itself is never post-processed actually, but wanted to keep the lifetimes simple since I Arc my `Vec<GlobFragment>` anyways